### PR TITLE
Move zeroes (Resolves #68)

### LINKED ISSUE #68: Move Zeroes to End of Array
## Statement
Given an integer array `nums`, write a function to move all `0`'s to the end of the array while maintaining the relative order of the non-zero elements.

### Example 1:

- **Input:**  `nums = [0,1,0,3,12]`  

- **Output:**  `[1,3,12,0,0]`

### Example 2:

- **Input:**  `nums = [0]`  

- **Output:**  `[0]`

### Constraints
- The operation must be performed **in-place** without making a copy of the array.

# Sources
[LeetCode](https://leetcode.com/problems/move-zeroes/description/)

### PR DESCRIPTION
**Pull request by mudassra-taskeen**
---
Name: Review solution of move_zeroes 
About: A template PR for code review of move_zeroes with a checklist
---

A function that moves all the zeros in a given integer array to the end, while keeping the relative order of the non-zero elements unchanged. The operation must be performed in-place, without using additional space for a copy of the array.

# Behavior

### Input:

- An integer array `(nums)`, which can contain:

    - Non-negative integers, including zeroes
    - The array length can vary

### Output:

- A modified array where:

    - All zeroes are moved to the end, while maintaining the relative order of the non-zero elements
    - The operation is done in-place without creating a copy of the array

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by jola-ds** [comment]
Hello, Mudassra. Good job, I will review your code.
It looks like py formmatting failed. It’s a problam with your second solution find_single _numbers.py. Could you resolve this? It also looks like you have 4 files in this PR. Maybe there’s a mix up with your branches? You also have both of your solutions in the same branch.

**Comment by mudassra-taskeen** [comment]
Hey Jola, Thanks for letting me know. I have removed those additional files from move_zeroes branch. Now you can review my code.

File: solutions/move_zeroes.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+Author: Mudassra Taskeen
+Date: 2025-01-08
+"""
+
+from typing import List
+
+
+def move_zeroes(nums: List[int]) -> None:
+    """
+    Move all 0's to the end of the input list in-place while maintaining the relative sequence of non-zero elements.
```
Comment (jola-ds) [inline solutions/move_zeroes.py]: Hi, Mudassara. Great job!

This is interesting. How can this docstring best reflect the behaviour of the function without directly describing the strategy? The function docstring is meant to describe the behaviour of the function, which is generally input and output.  Since this function only modifies the input, and returns None, it does not seem obvious how to describe the behaviour. Is it appropriate to say "Return None"? 

I think "Return a list of integers with all zeroes at the end" is not appropriate because the function does not return the list, it returns None. Can we bounce some ideas around?

File: solutions/move_zeroes.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        >>> nums = [0, 1, 0, 3, 12]
+        >>> move_zeroes(nums)
+        >>> nums
+        [1, 3, 12, 0, 0]
+
+        >>> nums = [0]
+        >>> move_zeroes(nums)
+        >>> nums
+        [0]
+    """
```
Comment (jola-ds) [inline solutions/move_zeroes.py:44]: This solution needs 3 doctests, could one more be added?

File: solutions/move_zeroes.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    assert isinstance(nums, List), "Input must be of type List"
+    assert all(isinstance(x, int) for x in nums), (
+        "All elements in the list must be integers"
+    )
+
+    # Initialize a pointer for the position of the next non-zero element
+    nz = 0
+
+    # Iterate through the list
+
```
Comment (jola-ds) [inline solutions/move_zeroes.py:54]: I think the comment can be just like the previous one and come right before the code. This space might not be necessary.

File: solutions/move_zeroes.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        [1, 3, 12, 0, 0]
+
+        >>> nums = [0]
+        >>> move_zeroes(nums)
+        >>> nums
+        [0]
+    """
+
+    # Defensive assertions
+    assert isinstance(nums, List), "Input must be of type List"
```
Comment (jola-ds) [inline solutions/move_zeroes.py]: The error should say list, not List, they are used differently and list is more appropriate in this case.

File: solutions/move_zeroes.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    # Initialize a pointer for the position of the next non-zero element
+    nz = 0
+
+    # Iterate through the list
+
+    for i in range(len(nums)):
+        if nums[i] != 0:
+            # Swap the current element with the element at nz
+
```
Comment (jola-ds) [inline solutions/move_zeroes.py:58]: This space might not be necessary as well

File: solutions/tests/test_move_zeroes.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    """Test cases for moveZeroes function."""
+
+    def test_standard_case(self):
+        """It should move all zeros to the end while maintaining order."""
+        nums = [0, 1, 0, 3, 12]
+        move_zeroes(nums)
+        self.assertEqual(nums, [1, 3, 12, 0, 0])
+
+    def test_single_zero(self):
+        """It should handle a single zero correctly."""
```
Comment (jola-ds) [inline solutions/tests/test_move_zeroes.py]: Is there a more specific description for this test? Something about the function leaving a single-zero list unchanged?

File: solutions/tests/test_move_zeroes.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        nums = [0]
+        move_zeroes(nums)
+        self.assertEqual(nums, [0])
+
+    def test_single_non_zero(self):
+        """It should handle a single non-zero element correctly."""
```
Comment (jola-ds) [inline solutions/tests/test_move_zeroes.py]: Is there a more specific description for this test? Something about the function leaving a single-element list unchanged?

File: solutions/tests/test_move_zeroes.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+Date: 2025-01-07
+Author: Mudassra Taskeen
+"""
+
+import unittest
+from ..move_zeroes import move_zeroes
+
+
+class TestMoveZeroes(unittest.TestCase):
+    """Test cases for moveZeroes function."""
```
Comment (jola-ds) [inline solutions/tests/test_move_zeroes.py]: I think this class docstring should refer to the actual name of the function, move_zeroes

File: solutions/move_zeroes.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    move_zeroes(nums): Rearranges the given list in-place as per the described behavior.
+
+Author: Mudassra Taskeen
+Date: 2025-01-08
+"""
+
+from typing import List
+
+
+def move_zeroes(nums: List[int]) -> None:
```
Comment (jola-ds) [inline solutions/move_zeroes.py:15]: May I ask why the function returns None, and not the modified list?

File: solutions/move_zeroes.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    """
+    Move all 0's to the end of the input list in-place while maintaining the relative sequence of non-zero elements.
+
+    Arguments:
+        nums (List[int]): The input list of integers.
+
+    Returns:
+        None: The function modifies the input list in-place.
```
Comment (jola-ds) [inline solutions/move_zeroes.py:24]: The return value may be updated depending on the result of our conversations.

**Review by jola-ds — CHANGES_REQUESTED** [review]
Well done, teammate! 

- The test suite is great, I would suggest adding more specific descriptions for some tests to make them even clearer.

- The function is well implemented as well. However, I'd like to have a conversation about the return value to make sure we're on the same page about the expected behaviour.

- It also looks like the function docstring could use an additional doctest, as the checklist specifies that we should have 3 or more passing doctests.

- Lastly, I think the assertion error message needs an update for clarity.

I understood your points and implemented them in my files. You can see those modifications, and let me know if I should alter anything else.

File: solutions/move_zeroes.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    """
+
+    # Defensive assertions
+    assert isinstance(nums, List), "Input must be of type List"
+    assert all(isinstance(x, int) for x in nums), (
+        "All elements in the list must be integers"
+    )
+
+    # Initialize a pointer for the position of the next non-zero element
+    nz = 0
```
Comment (jola-ds) [inline solutions/move_zeroes.py]: I think non_zero is a more direct and helpful variable name

**Review by jola-ds — CHANGES_REQUESTED** [review]

File: solutions/move_zeroes.py (hunk 1 above)
Comment (mudassra-taskeen) [reply solutions/move_zeroes.py]: I didn't understand your question. To the best of my knowledge, in this problem, we have to return the same list of integers with zeros moved at the end. When no new output is generated, the output would consequently be none. 
Here the function doesn't return a new list; only the position of zeroes is changed.

File: solutions/move_zeroes.py (hunk 10 above)
Comment (mudassra-taskeen) [reply solutions/move_zeroes.py:24]: How can return value rely on the result of our conversations?

File: solutions/move_zeroes.py (hunk 2 above)
Comment (mudassra-taskeen) [reply solutions/move_zeroes.py:44]: Thanks for informing me. I have added one more example in docstring

File: solutions/move_zeroes.py (hunk 3 above)
Comment (mudassra-taskeen) [reply solutions/move_zeroes.py:54]: Okay, I'll remove the additional space.

File: solutions/move_zeroes.py (hunk 4 above)
Comment (mudassra-taskeen) [reply solutions/move_zeroes.py]: I didn't notice that. Thanks for highlighting this mistake.

File: solutions/move_zeroes.py (hunk 5 above)
Comment (mudassra-taskeen) [reply solutions/move_zeroes.py:58]: I added some space to increase its readibility. That's why I did this to make the code more understandable.

File: solutions/tests/test_move_zeroes.py (hunk 6 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_move_zeroes.py]: You are right, but in that case I have to use the same words again. I guess I have already described the function so there is no need to add the same statement repeatedly. If you want me to change it then do let me know.

File: solutions/tests/test_move_zeroes.py (hunk 7 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_move_zeroes.py]: Hey Jola, I have already described the main purpose of this function in the file so I guess there is no need to repeat things in every test case.

File: solutions/move_zeroes.py (hunk 9 above)
Comment (mudassra-taskeen) [reply solutions/move_zeroes.py:15]: You are right here. I looked into it and thought it might not sound right. So, I made some modifications to the code. You can check that as well.

File: solutions/move_zeroes.py (hunk 11 above)
Comment (mudassra-taskeen) [reply solutions/move_zeroes.py]: Yeah, I have checked the checklist, and according to it, the variable names should describe the function properly. So, I removed the short names and wrote the precise variable names.

File: solutions/move_zeroes.py (hunk 1 above)
Comment (jola-ds) [reply solutions/move_zeroes.py]: So the function behaviour should be described in the docstring and that strategy should not be described in any docstring. I think this line describes the strategy a bit, so I was suggesting an improvement to that line. What do you think? 
Also, I understand better why it returns None 👍

File: solutions/move_zeroes.py (hunk 5 above)
Comment (jola-ds) [reply solutions/move_zeroes.py:58]: I understand and that’s good. However, spacing is meant to differentiate different stages of the implementation. I think this space is not necessary because the code is still in the same stage.

File: solutions/move_zeroes.py (hunk 10 above)
Comment (jola-ds) [reply solutions/move_zeroes.py:24]: It would need a change if you chose to update the return value or add a return call

File: solutions/move_zeroes.py (hunk 5 above)
Comment (mudassra-taskeen) [reply solutions/move_zeroes.py:58]: Your point is right.

File: solutions/move_zeroes.py (hunk 10 above)
Comment (mudassra-taskeen) [reply solutions/move_zeroes.py:24]: Oh, I see. Thanks for the clarification.

File: solutions/tests/test_move_zeroes.py (hunk 8 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_move_zeroes.py]: Thanks for highlighting this. I changed the name to move_zeroes.

File: solutions/move_zeroes.py (hunk 1 above)
Comment (mudassra-taskeen) [reply solutions/move_zeroes.py]: Alright, I did what you suggested. Truly grateful for considering every small change carefully.

**Review by fatima-malik99 — APPROVED** [review]
Hey, I've seen your code. Good job Mudassara. You did a great job. Going to merge it

**Review by jola-ds — APPROVED** [review]
